statsd/sender_resolving.go

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2012-2016 Eli Janssen
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file.
+
+package statsd
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+)
+
+// ResolvingSimpleSender provides a socket send interface that re-resolves and
+// reconnects.
+type ResolvingSimpleSender struct {
+	// underlying connection
+	c net.PacketConn
+	// resolved udp address
+	ra *net.UDPAddr
+	// unresolved addr
+	ro string
+	// addr changed
+	ac bool
+	// ra mutex
+	ramx sync.RWMutex
+	// interval counter
+	ticker *time.Ticker
+	// done chan
+	doneChan chan bool
+	running  bool
```

Comment:
This works, but i still think it's cleaner to tie `runmx` and `running` into the same struct and use an instance of that.  I think a copy/paste of `type AtomicBoolean struct{ flag uint32 }` from the link above could be useful.

Reply:
If you do that, make `AtomicBoolean` a private `atomicBoolean`

Reply:
You could also do the same with `ra *net.UDPAddr` to make the lock/unlock of their mutexes outside the core logic.

Reply:
🤔

Reply:
I don't think an atomic alone would provide the same guarantees needed here. The RWLock prevents a write lock from being held even after the check for running is done, providing coverage for the send call (while still allowing multiple read locks). 

Consider the case where one goroutine (A) checks running, and immediate after that, another (B) is calling Close and sets running. The socket send/close then becomes a racey. 

So while this mutex+bool construction is indeed a bit more awkward, it is providing a bit more behavioral safety vs simply protecting the running boolean itself.

Note: The release of the RLock could be done right after the send though, instead of as a defer. That would likely be a small optimization as currently defers do add a small amount of overhead. ref: https://medium.com/i0exception/runtime-overhead-of-using-defer-in-go-7140d5c40e32